leader/src/jerigon.rs

```diff
@@ -10,45 +10,85 @@ use ethereum_types::H256;
 use paladin::runtime::Runtime;
 use proof_gen::types::PlonkyProofIntern;
 
+use crate::cli::BlockNumbers;
+
 /// The main function for the jerigon mode.
 pub(crate) async fn jerigon_main(
     runtime: Runtime,
     rpc_url: &str,
-    block_numbers_range: [u64; 2],
+    block_numbers: BlockNumbers,
     checkpoint_block_number: u64,
     previous: Option<PlonkyProofIntern>,
     proof_output_dir_opt: Option<PathBuf>,
 ) -> Result<()> {
-    let mut block_number = block_numbers_range[0];
     let mut previous = previous;
     let mut block_hash_cache: HashMap<u64, H256> = HashMap::new();
-    while block_number <= block_numbers_range[1] {
-        let prover_input = rpc::fetch_prover_input(rpc::FetchProverInputRequest {
-            rpc_url,
-            block_number,
-            checkpoint_block_number,
-            block_hash_cache: &mut block_hash_cache,
-        })
-        .await?;
-
-        let proof = prover_input.prove(&runtime, previous).await?;
 
-        let proof_json = serde_json::to_vec(&proof.intern)?;
-        write_proof(
-            proof_json,
-            proof_output_dir_opt
-                .as_ref()
-                .map(|p| p.join(format!("b{}.zkproof", block_number))),
-        )?;
-
-        previous = Some(proof.intern);
-        block_number += 1;
+    match block_numbers {
+        BlockNumbers::Single(block_number) => {
+            process_block(
+                &runtime,
+                rpc_url,
+                block_number,
+                checkpoint_block_number,
+                previous,
+                &proof_output_dir_opt,
+                &mut block_hash_cache,
+            )
+            .await?;
+        }
+        BlockNumbers::RangeInclusive(block_numbers_range) => {
+            // Code to execute if BlockNumbers is RangeInclusive
+            for block_number in block_numbers_range {
+                previous = process_block(
+                    &runtime,
+                    rpc_url,
+                    block_number,
+                    checkpoint_block_number,
+                    previous,
+                    &proof_output_dir_opt,
+                    &mut block_hash_cache,
+                )
+                .await?;
+            }
+        }
```

Comment:
Sorry, one more nit! Wdyt about making the `match` block return a range for both cases that gets processed by a loop? Then we can remove the duplicated call to `process_block` in each arm. Optional, so I'll let you make the final call.

Reply:
good idea. thx :)